### PR TITLE
Create an edit aligned text mode - visual editor - step 5

### PR DESCRIPTION
**Pull request by irina060981**
For the issue https://github.com/alpheios-project/alignment-editor-new/issues/39

- added undo/redo, history for tokens editor
- some refactoring for history, history step to use the same code for both histories - alignment groups and tokens edits
- some refactoring for alignment and alignment-group for separating code to actions class
- some unit tests and descriptions

@balmas, I created steps constants inside `HistoryStep` (not inside `AlignmentGroupStep` and `TokensEditStep`), because babel imports classes asynchronously and this leads to the fact that `HistoryStep` is imported after children classes and jest throws an error.

I added not all descriptions and tests - only those I decided to choose as most important due to save time for the other work.

P.S. Bridget, about deployment - these changes from the PR is uploaded to the production (before PR is merged) - https://alpheios-misc-dev.s3.us-east-2.amazonaws.com/alignment-editor/index.html
Is it correct?

### CODE REVIEW

**Comment by balmas** [comment]
>P.S. Bridget, about deployment - these changes from the PR is uploaded to the production (before PR is merged) - https://alpheios-misc-dev.s3.us-east-2.amazonaws.com/alignment-editor/index.html
Is it correct?

Yes -- that environment is now the 'dev' environment. "Production" is at https://alignment.alpheios.net . Changes should only be deployed there upon push/merge to master

**Review by balmas — APPROVED** [review]
I think this is a good step forward with the handling of the history
I didn't review the code super closely, especially the pieces that looked like they were just moved from one place to another
I think you could do a search and replace on 'prev' and 'next' to make sure to use the constants everywhere (and in some cases you can also simplify logic that tests on them).

File: src/lib/data/actions/tokens-edit-actions.js [hunk 1]
```diff
@@ -0,0 +1,516 @@
...
+   *
+   * @param {Token} token
+   * @param {String} tokenType - prev/next
+   * @returns {Object}
+   *          {Segment} segment
+   *          {Number} tokenIndex
+   *          {String} position - prev/next
+   *          {Token} tokenResult
+   */
+  getToken (token, tokenType) {
```
[the remark this answers is not in the document]
Comment (irina060981) [reply src/lib/data/actions/tokens-edit-actions.js:58]: right!

File: src/lib/data/actions/tokens-edit-actions.js [hunk 2]
```diff
@@ -0,0 +1,516 @@
...
+    return token.update({ word, idWord: newIdWord })
+  }
+
+  /**
+   * @param {Token} token - token for update
+   * @param {String} direction - left/right
+   * @returns {Boolean}
+   */
+  mergeToken (token, direction) {
+    const { segment, tokenIndex, tokenResult, position } = (direction === HistoryStep.directions.PREV) ? this.getToken(token, 'prev') : this.getToken(token, 'next')
```
[the remark this answers is not in the document]
Comment (irina060981) [reply src/lib/data/actions/tokens-edit-actions.js]: fixed

File: src/lib/data/actions/tokens-edit-actions.js [hunk 3]
```diff
@@ -0,0 +1,516 @@
...
+
+  /**
+   * Moves the token to the next/previous segment
+   * @param {Token} token
+   * @param {HistoryStep.directions} direction
+   * @returns {Boolean}
+   */
+  moveToSegment (token, direction) {
+    const segment = this.getSegmentByToken(token)
+    const newSegment = (direction === HistoryStep.directions.PREV) ? this.getSegmentByToken(token, 'prev') : this.getSegmentByToken(token, 'next')
```
[the remark this answers is not in the document]
Comment (irina060981) [reply src/lib/data/actions/tokens-edit-actions.js]: Don't know why I didn't guess ti it myself

File: src/lib/data/alignment.js [hunk 4]
```diff
@@ -672,97 +693,17 @@ export default class Alignment {
    * @returns {Boolean}
    */
   mergeToken (token, direction) {
-    const { segment, tokenIndex, tokenMergeTo, position } = (direction === TokensEditController.direction.PREV) ? this.getPrevToken(token) : this.getNextToken(token)
+    const { tokenResult } = (direction === HistoryStep.directions.PREV) ? this.tokensEditActions.getToken(token, 'prev') : this.tokensEditActions.getToken(token, 'next')
```
[the remark this answers is not in the document]
Comment (irina060981) [reply src/lib/data/alignment.js]: fixed and all similiar

File: src/locales/en-us/messages-tokens-edit-editor.json [hunk 5]
```diff
@@ -83,6 +83,26 @@
...
+      "description": "A label for action menu buttons",
+      "component": "ActionsMenuTokensEditor"
+    },
+    "ACTIONS_REDO_TITLE": {
+      "message": "Redo",
+      "description": "A label for action menu buttons",
+      "component": "ActionsMenuTokensEditor"
+    },
+    "TOKENS_EDIT_UNDO_ERROR": {
+      "message": "There is no steps to be undone for tokens editor.",
```
[the remark this answers is not in the document]
Comment (irina060981) [reply src/locales/en-us/messages-tokens-edit-editor.json]: fixed

File: src/locales/en-us/messages-tokens-edit-editor.json [hunk 6]
```diff
@@ -83,6 +83,26 @@
...
+      "description": "An error inside tokens edit history workflow",
+      "component": "Alignment"
+    },
+    "TOKENS_EDIT_REDO_ERROR": {
+      "message": "There is no steps to be redone for tokens editor.",
```
[the remark this answers is not in the document]
Comment (irina060981) [reply src/locales/en-us/messages-tokens-edit-editor.json]: fixed

**Pull request merged**